packaging/common/substitute.pl

```diff
@@ -125,92 +266,75 @@ sub fix_pkg_rev {
-            # Releases can have 3 fields on Debian.  we want the first 2.
-            $release =~ s/(\d+)\.(\d+).*/$1$2/;
-        }
-        $replacement_strings{"%%DEB_REL%%"} = $release;
 	$replacement_strings{"%%DATE%%"} = get_date($replacement_strings{"%%DATE%%"});
-	# 32bit should use bitrock perl, while 64bit should use builtin.  we
-	# live on the edge and assume it's there.
-	my $arch = get_arch();
-	if ( $arch eq "x86_64" ) {
-		$replacement_strings{"%%PERL%%"} = $^X;
-	}
-	else {
+	if ( get_arch() eq "x86_64" ) {
 		$replacement_strings{"%%PERL%%"} = "/opt/zmanda/amanda/perl/bin/perl";
 	}
 }
-elsif ( $pkg_type eq "rpm" ){
+
+if ( $pkg_type eq "rpm" ){
 	%replacement_strings = ( %replacement_strings_rpm,
 				 %replacement_strings_common );
         $replacement_strings{"%%PKG_REV%%"} =
             fix_pkg_rev($replacement_strings{"%%PKG_REV%%"}, "rpm");
 	$replacement_strings{"%%DATE%%"} = get_date($replacement_strings{"%%DATE%%"});
 }
-else {
+
+if ( $pkg_type eq "sun" ){
 	%replacement_strings = ( %replacement_strings_sun,
 				 %replacement_strings_common );
         $replacement_strings{"%%PKG_REV%%"} =
             fix_pkg_rev($replacement_strings{"%%PKG_REV%%"}, "sun");
 	$replacement_strings{"%%DATE%%"} = get_date($replacement_strings{"%%DATE%%"});
+
 	my $arch = get_arch();
-	if ( $arch eq "sun4u" ) {
-	    $replacement_strings{"%%ARCH%%"} = "sparc";
-	}
-	elsif ( $arch eq "i86pc" ) {
-	    $replacement_strings{"%%ARCH%%"} = "intel";
-	}
-	else {
-	    die "Unknown solaris platform!";
-	}
+        $replacement_strings{"%%ARCH%%"} = "sparc"
+            if ( $arch eq "sun4u" );
+        $replacement_strings{"%%ARCH%%"} = "intel"
+            if ( $arch eq "i86pc" );
+        die "Unknown solaris platform!"
+            if ( ! $replacement_strings{"%%ARCH%%"} );
 }
 
 # Make a hash of tags and the contents of replacement files
 my %replacement_data;
 while (my ($tag, $filename) = each %replacement_filenames) {
 	open(my $file, "<", $filename) or die "could not read \"$filename\": $!";
-	$replacement_data{$tag} = join "", <$file>;
+        {
+            local $/;  # remove line endings as barrier
+            $replacement_data{$tag} = join "", <$file>;
+        }
 	close($file);
 }
 open my $src, "<", $ARGV[0] or die "could not read $ARGV[0]: $!";
 open my $dst, ">", $ARGV[1] or die "could not write $ARGV[1]: $!";
 select $dst;
+
+my $line;
+
 while (<$src>) {
-	chomp;
-	# check for tags, using non greedy matching
-	if ( m/(%%.+?%%)/ ) {
-		# Data replaces the line
-		if ( defined($replacement_data{$1})) {
-			print $replacement_data{$1};
-		} 
-		# strings just replace the tag.
-		elsif ( defined($replacement_strings{$1})) {
-			s/(%%.+?%%)/$replacement_strings{$1}/g;
-			print "$_\n";
-		}
-	}
-	else {
-		# If we got here, print the line unmolested
-		print "$_\n";
+    # check for tags, using non greedy matching
+    #
+    $line = $_;
+    SUBST:
+    # for all tags found in the line
+    while ( m/(%%\w+?%%)/g ) {
```

Comment:
we *may* have multiple tags in one line.. so the loop is needed and each distinct $tag needs matching once.

assumption is (of course) that tags found later on are not subverted by the data-contents of the earlier ones on the line.